keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/LaunchConfiguration.kt

```diff
@@ -21,9 +21,40 @@ data class LaunchConfiguration(
   val imageId: String,
   val appVersion: String,
   val instanceType: String,
-  val ebsOptimized: Boolean,
+  val ebsOptimized: Boolean = false,
   val iamRole: String,
   val keyPair: String,
   val instanceMonitoring: Boolean = false,
   val ramdiskId: String? = null
-)
+) {
+  fun toLaunchConfigurationSpec(account: String, application: String, omitDefaults: Boolean = false) =
+    ClusterSpec.LaunchConfigurationSpec(
+      instanceType = instanceType,
+      ebsOptimized = if (omitDefaults && !ebsOptimized) {
+        null
+      } else {
+        ebsOptimized
+      },
+      iamRole = if (omitDefaults && iamRole == defaultIamRoleForApp(application)) {
+        null
+      } else {
+        iamRole
+      },
+      keyPair = if (omitDefaults && keyPair == defaultKeyPairForAccount(account)) {
+        null
+      } else {
+        keyPair
+      },
+      instanceMonitoring = if (omitDefaults && !instanceMonitoring) {
+        null
+      } else {
+        instanceMonitoring
+      },
+      ramdiskId = ramdiskId
+    )
+
+  companion object {
+    fun defaultKeyPairForAccount(account: String) = "nf-$account-keypair-a"
```

Comment:
is there a way we can reach out to clouddriver to ask it for the default keypair for an account rather than assuming it's going to look like this? IIRC clouddriver is aware of the appropriate keypair to use.

Reply:
The [keyPair controller](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/KeyPairController.groovy) is what you should use!

Reply:
@emjburns That API just returns the entire list of known key pairs. How do I tell which one is the default for an account? The JSON for each key pair has just these fields:
```json
  {
    "account": "test",
    "region": "us-west-2",
    "keyName": "<redacted>",
    "keyFingerprint": "<redacted>"
  }
```

Reply:
Aha! Good call. It looks like the [credentials controller](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-web/src/main/groovy/com/netflix/spinnaker/clouddriver/controllers/CredentialsController.groovy#L54) will return this information (`GET /credentials/ACCOUNT`) as the value `defaultKeyPair`. It uses some templating for the region, so I think what we should do is cache the default keypairs for all accounts and regions so we can easily look this up.

Reply:
Just adding a note here based on peer review with @emjburns: it looks like we actually have two different naming conventions for key pair names. One is `nf-$account-keypair-a` like I defined above, which we assume takes effect whenever there's not a `defaultKeyPair` specified for that account in the CloudDriver config; the other is `nf-keypair-${account}-${region}`, which many of the accounts in our particular config follow.

In the former case, since the name will match the convention regardless of region, it would be obvious to omit from the export if it also matches what CloudDriver reports as the default key pair name. In the latter, it's a little trickier, but we discussed possibly adding a new type of `Resolver` to query the CloudDriver cache and have some logic to accommodate for both those patterns. However, I'm still unsure how to adequately use the resolver pattern in the export use case since, as I pointed out before, resolvers will actually _apply_ defaults and _add_ detail, whereas the point of this PR is to _remove_ detail from the reverse-engineered specs...

Please share your thoughts.

Reply:
It looks like deck does nothing but run `const defaultKeyPair = acct.defaultKeyPair.replace('{{region}}', command.region)` to handle the `{{region}}` expansion for some of the `defaultKeyPair` values returned by: https://clouddriver/credentials?expand=true. It seems reasonable to directly replicate that behavior and build a `defaultKeypairForAccount(account: String): String` method somewhere that utilizes a local in-memory cache of the account credential data. I wouldn't try to fit it into the `Resolver` interface for the reasons you gave. Could just be a member of `MemoryCloudDriverCache`.

Reply:
@emjburns @asher I pushed a commit refactoring the way the default key pair was handled during export to use clouddriver data as discussed + other changes for consistency. Let me know if it looks decent.

Reply:
I'd vote for deferring any more complex handling of keypairs to a further PR that addresses that issue in isolation.